### ISSUE TITLE
Add checks for relative input shapes inside bindings.

### ISSUE THREAD

**Issue by NippunSharma** [mlpack/mlpack#2820]
### What is the desired addition or change?

A parameter check should be added to all bindings that checks for the correct relative input shapes. For example the training data and training labels must have the same number of samples, testing data and training data must have the same number of features etc.

### What is the motivation for this feature?

Refer to #2818 for more information and how to reproduce the bug.

### If applicable, describe how this feature would be implemented.

- As discussed with @rcurtin the checks should be implemented inside `src/mlpack/core/util/param_check.hpp` and then we should call these checks from the bindings.

- You can take a look at how `RequireOnlyOnePassed` is used the `_main.cpp` files for each method in the `mlpackMain()` function and is implemented in `src/mlpack/core/util/param_check_impl.hpp`. These checks would also be implemented and used similarly.

- We need to check each binding (listed below) separately and add these checks if they are not already there. For example in `Linear SVM` same number of columns between training data and labels are already checked [here](https://github.com/mlpack/mlpack/blob/bf837f950b728bf08b80ccc696c6bc0d296e54af/src/mlpack/methods/linear_svm/linear_svm_main.cpp#L281).

- [ ] adaboost
- [ ] decision_stump
- [ ] decision_tree
- [ ] hoeffding_tree
- [ ] linear_svm
- [ ] logistic_regression
- [ ] nbc
- [ ] perceptron
- [ ] random_forest
- [ ] softmax_regression
- [ ] bayesian_linear_regression
- [ ] lars
- [ ] linear_regression
- [ ] dbscan
- [ ] gmm_train
- [ ] gmm_generate
- [ ] gmm_probability
- [ ] kmeans
- [ ] mean_shift
- [ ] approx_kfn
- [ ] emst
- [ ] fastmks
- [ ] lsh
- [ ] knn
- [ ] kfn
- [ ] krann
- [ ] preprocess_split
- [ ] preprocess_binarize
- [ ] preprocess_describe
- [ ] preprocess_scale
- [ ] preprocess_one_hot_encoding
- [ ] image_converter
- [ ] cf
- [ ] det
- [ ] hmm_train
- [ ] hmm_loglik
- [ ] hmm_viterbi
- [ ] hmm_generate
- [ ] kde
- [ ] nmf
- [ ] kernel_pca
- [ ] lmnn
- [ ] local_coordinate_coding
- [ ] nca
- [ ] pca
- [ ] radical
- [ ] sparse_coding

Note that not all of the bindings written above will require these checks because for some they would be there already. So it would be better to check this before changing any code (please comment below after checking so that I can update the description to indicate where the checks are not required).

### Additional information?

If you can think of a better way to implement it then you can comment below and we can discuss it.

Update: Also replace the occurences where we are comparing sizes of matrices with utilities created in #2370.

**Comment by NippunSharma** [comment]
I have changed the description above. Please check it again if you have already started working on this. Sorry for any inconvenience caused.

**Comment by onikolskyy** [comment]
Hey, I would check 

- [ ] adaboost

- [ ] decision_stump

- [ ] decision_tree

- [ ] hoeffding_tree

if these points are still free. I consider it a good starting point to dive into the project. Would take me about a week.

**Comment by NippunSharma** [comment]
Hi @onikolskyy , sure feel free. Remember that you will also have to implement the checking functions inside `param_check.hpp` and then add to the bindings (if we are not already checking for the same conditions in some way or another). You can post any questions you have over here.

**Comment by shubhamabhang77** [comment]
Is there anyone working on this issue??
I would like to work this to get started contributing.

**Comment by shubhamabhang77** [comment]
Is there anyone working on this issue??
I would like to work this to get started contributing. 
We have add write declaration in ```param_check.hpp``` and add the definition in ```param_check_impl.hpp```, right?

**Comment by onikolskyy** [comment]
@shubhamabhang77 mate, I am working on the adaboost bug #2818 right now , would you mind taking one of the other points?
I will report my status later today/tomorrow

**Comment by onikolskyy** [comment]
@NippunSharma I have an issue with setting up a dev environements for the bindings.
I have uploaded a question, perhaps there is a quick answer?  #2843

**Comment by shubhamabhang77** [comment]
yeah yeah sure

**Comment by onikolskyy** [comment]
Hey guys, 
I would be happy to make my first commit with a fix of  #2818.
However, I get a `permission denied` for my `github` acc when I try to push a new branch with the commit.
Do I need some extra rights or does the contribution pipeline work differently?

**Comment by NippunSharma** [comment]
You cannot push directly to the mlpack master, you will not have the access. You will have to create a fork of this repo and then push to your fork and then create a pull request. You can look up on the net, the process is very easy and will not take much time.

**Comment by roshanjain77** [comment]
Hi, Is there anyone currently working on this issue?
I would like to work on the above issue.

**Comment by rcurtin** [comment]
Two updates relevant to this issue:

 - #2370 contains some nice utilities for checking the size of a matrix.  We should probably use those in whatever we build here.
 - [This comment](https://github.com/mlpack/mlpack/pull/2370#pullrequestreview-436211692) tries to find all the existing places in the codebase where we are comparing sizes of matrices.  All those usages should be replaced with the utilities being written here (if it is in a binding), or with the functions from #2370 (if not in a binding).  Before closing this issue, we should be sure to check that that grep comes up empty. :)

**Comment by Rishabhc711** [comment]
I would like to work on this issue. I have looked through the comments and understood the problem. Where in the codebase should I look for functions that might need to check for the dimensions of matrices?

As in where should I start off and what approach should I follow?

**Comment by NippunSharma** [comment]
Keep open.

**Comment by PranavReddyP16** [comment]
Hey, @rcurtin @NippunSharma I assumed that this hasn't been fixed since the last comment is from a month ago so I thought I'd take a look at it. I was thinking of replacing all the occurrences of size and dimensionality checks using the functions added in #2370. I was taking a look at logistic regression and there was an instance of code that seemed similar to the function but can't be replaced directly by it - [Logistic Regression Code instance](https://github.com/mlpack/mlpack/blob/e39065c32fc67bba2ef4bd4b44f75cef20816a16/src/mlpack/methods/logistic_regression/logistic_regression_main.cpp#L353). Would it be better to rewrite it slightly or keep it as it is? 
Thanks!

**Comment by NippunSharma** [comment]
Hey @PranavReddyP16, thanks for showing interest in this issue. As you can read from the description we only need to include checks where they are not present. There is no need to replace the "if" conditions and other occurrences, at least for this issue. For better understanding, you can take a look at #2818, which is the actual issue we are facing. Let me know if you have any other doubts 👍

**Comment by PranavReddyP16** [comment]
> Two updates relevant to this issue:
> 
> * #2370 contains some nice utilities for checking the size of a matrix.  We should probably use those in whatever we build here.
> * [This comment](https://github.com/mlpack/mlpack/pull/2370#pullrequestreview-436211692) tries to find all the existing places in the codebase where we are comparing sizes of matrices.  All those usages should be replaced with the utilities being written here (if it is in a binding), or with the functions from #2370 (if not in a binding).  Before closing this issue, we should be sure to check that that grep comes up empty. :)

Hey, @NippunSharma thanks for the clarification. I was a little confused because of the second point of this comment from Ryan and I thought we had to replace if statements _and_ add checks where needed. I'll get to work on adding check functions if they're not present. Just to confirm, I can use the functions implemented in #2370 in the `(method)_main.cpp` files in the `mlpackMain()` function am I right? Thanks again!

**Comment by NippunSharma** [comment]
Ya, you are right. I am sorry for the confusion here. I forgot to update the issue description, I guess that is why it slipped my mind. I have added that now, thanks for reminding 👍.  Just to clarify again, you will have to replace all other occurrences along with the checks. Yes you have to use functions implemented in #2370.You can create a PR and we can discuss any other doubts you have over there 👍

**Comment by PranavReddyP16** [comment]
Cool, thanks a lot for clearing that up. I'll get started on it.

**Comment by ayushi19031** [comment]
I would like to work on this

**Comment by NippunSharma** [comment]
@ayushi19031, go ahead. I think @onikolskyy @PranavReddyP16 were working on this, best to ask them first 👍

**Comment by onikolskyy** [comment]
HI @ayushi19031, I haven't worked on this since my last contribution. Please feel free to continue

**Comment by VedangAsgaonkar** [comment]
Is this issue still open?

**Comment by rcurtin** [comment]
@VedangAsgaonkar yes, as you can see, the issue is not closed, so it is still open. :smile: 

Please feel free to take a look at some of the PRs that have already been done for other bindings (you can look through the list of linked PRs), then pick a binding and open a PR for it.  I'd recommend smaller PRs, one per binding, which will make it easier to get them merged.  I think the description that @NippunSharma left above is sufficient to get started. :+1:

**Comment by eshaanagarwal** [comment]
Hi ! I would like to take up PCA first and then move on to preprocess_split and other bindings.

**Comment by AdarshSantoria** [comment]
I would like to work on this issue.

**Comment by rcurtin** [comment]
@AdarshSantoria feel free, if you open a PR it will be reviewed when possible. :+1:

**Comment by shubhksingh** [comment]
where am i suppose to implement param check as previous merged pr implemented parm check in linearregression.cpp  file itself
and not in param_check.hpp

**Comment by lumi232** [comment]
Hi, I have replaced all the if checks in the Linear Regression method with the utility introduced in #2370. I have made a PR looking forward to it being merged. I also plan to do this with other methods after this one is merged successfully.

**Comment by sukjingitsit** [comment]
Hello, my name is Sukrit and I’m a 2nd year undergrad at IIT Roorkee, India. I have only recently started with open source contributions, and I identified this as a good first issue for me to tackle. I would like to address this issue as specifically for PCA, NCA, Kernel-PCA and Perceptrons as a starting point. I am currently going through the codebase of mlpack to get an understanding of the code and dealing with the issue and will update once I start working on the coding part (expected in 1-2 days). So far, as mentioned in the above discussion, I believe using methods from size_checks.hpp will be the way forward.
